netlogo-gui/src/main/ide/CodeCompletionPopup.scala

```diff
@@ -87,48 +91,66 @@ case class CodeCompletionPopup(compiler: CompilerServices) {
 
   /**
     * Updates the list of the suggestion box and makes it visible.
     */
-  def updatePopup(): Unit = {
-    println("I was called")
+  def updatePopup(docEventOption: Option[DocumentEvent]): Unit = {
     for{eA <- editorArea} {
-      if (isPopupEnabled) {
-        val token = compiler.getTokenAtPosition(eA.getText(), eA.getCaretPosition)
-        var list = Seq[String]()
-        // word only contains the current token till the cursor position
-        val word = if (token != null) token.text.substring(0, eA.getCaretPosition - token.start) else null
-        // popup not to be diplayed after user hits the enter key
-        if (word != null && !token.text.equals(lastSuggested)) {
-          list = autoSuggest.getSuggestions(word)
-          dlm.removeAllElements()
-          list.foreach(dlm.addElement(_))
-          if (!list.isEmpty) {
-            lastSuggested = ""
-            window.validate()
-            window.setVisible(true)
-            // Required to keep the caret in the editorArea visible
-            eA.getCaret.setVisible(true)
-            suggestionDisplaylist.setSelectedIndex(0)
-          } else {
-            isPopupEnabled = false
+      if (isPopupEnabled ) {
+        var position = eA.getCaretPosition
+        for{docEvent <- docEventOption} {
+          docEvent.getType match {
+            // Trying to fix the behavior of getCaretPosition which is not returning the correct position
+            case DocumentEvent.EventType.INSERT => position += 1
+            case _ =>
           }
         }
-        if (list.isEmpty || token == null) {
-          window.setVisible(false)
+        val tokenOption = getTokenTillPosition(eA.getText(), position)
+        var list = Seq[String]()
+        tokenOption match {
+          case Some(token) =>
+            // word only contains the current token till the cursor position
+            val word = token.text.substring(0, position - token.start)
+            // popup not to be diplayed after user hits the enter key
+            if (!token.text.equals(lastSuggested)) {
+              list = autoSuggest.getSuggestions(word)
+              dlm.removeAllElements()
+              list.foreach(dlm.addElement(_))
+              if (!list.isEmpty) {
+                lastSuggested = ""
+                window.validate()
+                window.setVisible(true)
+                // Required to keep the caret in the editorArea visible
+                eA.getCaret.setVisible(true)
+                suggestionDisplaylist.setSelectedIndex(0)
+              } else {
+                isPopupEnabled = false
+              }
+            }
+            if (list.isEmpty || token == null) {
+              window.setVisible(false)
+            }
+          case None => window.setVisible(false)
         }
       }
     }
   }
-  def isVisible() = window.isVisible
+  def getTokenTillPosition(source: String, position: Int): Option[Token] = {
+    val iterator = Tokenizer.tokenizeString(source)
```

Comment:
Instead of using `Tokenizer`, use `org.nlogo.core.Femto.scalaSingleton[TokenizerInterface]("org.nlogo.lex.Tokenizer")`. Why go through all of that just to use the same object in the end? Because it ensures that this class depends on the stable published interface `org.nlogo.core.TokenizerInterface` instead of `org.nlogo.lex.Tokenizer`, which might change for any number of reasons unrelated to this class.